privval/rs_client.go

```diff
@@ -0,0 +1,236 @@
+package privval
+
+import (
+	"fmt"
+	"github.com/tendermint/tendermint/crypto"
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/types"
+	"net"
+	"sync"
+	"time"
+)
+
+// RemoteSignerClient implements PrivValidator, it uses a socket to request signatures
+// from an external process.
+type RemoteSignerClient struct {
+	cmn.BaseService
+
+	connHeartbeat time.Duration
+
+	conn       net.Conn
+	lock       sync.Mutex
```

Comment:
What is the reason for the lock? `net.Conn` is threadsafe.

Reply:
The idea was that if the KMS does asynchronous receive the following could happen:

TM sends a sign request.
TM send a getpubkey request.

The KMS handles both but for whichever reason it finishes the getpubkey request first.

Then the sign method would receive the packet for the getpubkey then since we have no package IDs etc. 

As long as the KMS side does proper synchronizing this is not an issue but we should take care of this here.

Or do I have misconceptions about how this works ?